codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,329 +16,514 @@
+    public final void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
         if (!(evt instanceof UpgradeEvent)) {
             super.userEventTriggered(ctx, evt);
             return;
         }
 
         UpgradeEvent upgrade = (UpgradeEvent) evt;
-        ctx.fireUserEventTriggered(upgrade.retain());
+        onUpgradeEvent(ctx, upgrade.retain());
         try {
-            Http2Stream stream = http2Handler.connection().stream(Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
+            Http2Stream stream = connection().stream(Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
             // TODO: improve handler/stream lifecycle so that stream isn't active before handler added.
             // The stream was already made active, but ctx may have been null so it wasn't initialized.
             // https://github.com/netty/netty/issues/4942
             new ConnectionListener().onStreamActive(stream);
             upgrade.upgradeRequest().headers().setInt(
                     HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
-            new InboundHttpToHttp2Adapter(http2Handler.connection(), http2Handler.decoder().frameListener())
+            new InboundHttpToHttp2Adapter(connection(), decoder().frameListener())
                     .channelRead(ctx, upgrade.upgradeRequest().retain());
         } finally {
             upgrade.release();
         }
     }
 
-    // Override this to signal it will never throw an exception.
-    @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        ctx.fireExceptionCaught(cause);
-    }
-
     /**
      * Processes all {@link Http2Frame}s. {@link Http2StreamFrame}s may only originate in child
      * streams.
      */
     @Override
-    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
-        try {
-            if (msg instanceof Http2WindowUpdateFrame) {
-                Http2WindowUpdateFrame frame = (Http2WindowUpdateFrame) msg;
-                consumeBytes(frame.streamId(), frame.windowSizeIncrement(), promise);
-            } else if (msg instanceof Http2StreamFrame) {
-                writeStreamFrame((Http2StreamFrame) msg, promise);
-            } else if (msg instanceof Http2GoAwayFrame) {
-                writeGoAwayFrame((Http2GoAwayFrame) msg, promise);
-            } else {
-                throw new UnsupportedMessageTypeException(msg);
-            }
-        } finally {
+    public final void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        write0(ctx, msg, promise);
+    }
+
+    // Allow to override for testing
+    void write0(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        if (msg instanceof Http2DataFrame) {
+            Http2DataFrame dataFrame = (Http2DataFrame) msg;
+            encoder().writeData(ctx, dataFrame.stream().id(), dataFrame.content(),
```

Comment:
followup is fine ... can you create an issue and link it here? this caused an issue in the example in this PR so I think its worth tracking, as it seems likely others may make the same mistake.